src/basic/macro.h

```diff
@@ -514,4 +514,13 @@ static inline int __coverity_check__(int condition) {
                 return r < 0 ? EXIT_FAILURE : EXIT_SUCCESS;             \
         }
 
+/* Zero is mapped to EXIT_SUCCESS, and both negative and positive values
+ * are mapped to EXIT_FAILURE.
+ * Note: this means "true" maps to EXIT_FAILURE. */
+#define DEFINE_MAIN_FUNCTION_WITH_POSITIVE_FAILURE(impl)                \
+        int main(int argc, char *argv[]) {                              \
+                int r = impl(argc, argv);                               \
+                return r != 0 ? EXIT_FAILURE : EXIT_SUCCESS;            \
```

Comment:
hmm, wouldn't it be better to do a single macro, but with the semantics of "map 0…255 to itself, everything else to EXIT_FAILURE"? this would mean, all negative errors would automatically become failure, but if some function wants to return something else it can just do so 1:1, without any tricks and hidden mappings.

I think that would cover all cases in one, no?

Reply:
I first wrote it this way, but then it's necessary to make sure that `r` is not positive. Many our functions return `1`, and it is usual to propagate the return value through many layers, so this can be hard. And of course somebody could "harmlessly" modify one of those functions somewhere deep in the call stack. Essentially, in our coding convention, non-negatives are generally equivalent, but here they need special handling. Doing `r < 0 ? EXIT_FAILURE : EXIT_SUCCESS` was very common for exactly this reason, and trying to shoehorn both types into the same function would be very error-prone IMHO.

Reply:
hmm, I am still confused by this one though. In my head this one is kinda strange, because weirdly inverted... i.e. i.e. typically internally we say "0" means "false" and "> 0" means "true". On UNIX program exit codes are the opposite of course, "0" means "success" and "> 0" means "failure". Now, it appears to me that this macro should easy remove this discrepancy (and hence do the opposite of what you do above, i.e. it should map > 0 to 0 and 0 to 1), or leave the return code in the valid range as it is and propagate it unmodified (which is also different from your patch, as you map >= 2 to 1).  

Hence, I am wondering, what did I miss? Why did you choose to go for a definition that keeps the "inverted" meaning, but still maps > 1 to 1?

Reply:
So there's two questions:
>  leave the return code in the valid range as it is and propagate it unmodified

I did that because that's what the code I was replacing did. (For most executables) we don't define return codes other than 0 and 1. So just propagating the the return value would "create" new values which weren't there before. If it ever becomes useful, we can of course add that, but so far this wouldn't be used.

> typically internally we say "0" means "false" and "> 0" means "true"

Dunno, both options would work. So far we have:
```C
boot-check-no-failures.c:        return n > 0;
detect-virt.c:        return r == VIRTUALIZATION_NONE;
fsck.c:        return exit_status & (FSCK_SYSTEM_SHOULD_REBOOT | FSCK_ERRORS_LEFT_UNCORRECTED);
inhibit.c:                return wait_for_terminate_and_check(argv[optind], pid, WAIT_LOG);
                             return print_inhibitors(bus);
```
(with `DEFINE_MAIN_FUNCTION_WITH_POSITIVE_FAILURE(run);`)
This would become:
```C
boot-check-no-failures.c:        return n == 0;
detect-virt.c:        return r != VIRTUALIZATION_NONE;
fsck.c:        return (exit_status & (FSCK_SYSTEM_SHOULD_REBOOT | FSCK_ERRORS_LEFT_UNCORRECTED)) == 0;
inhibit.c:                return wait_for_terminate_and_check(argv[optind], pid, WAIT_LOG) == 0;
                             return print_inhibitors(bus) == 0;
```
(with `DEFINE_MAIN_FUNCTION_WITH_POSITIVE_SUCCESS(run);`)

If you think this is better, I can adjust the PR.